### PR TITLE
Bump minimum CMake version up to 3.16

### PR DESCRIPTION
**Pull request by hebasto**
Considering the time when the new CMake-based build system will be available for users, it seems reasonable to drop support
for CMake versions older than 3.16.

Here are benefits of such version bumping:
- code simplification (see individual commits)
- the `cmake --build` tool gained `--verbose` and `-v` options to specify verbose build output
- the `cmake-gui` dialog gained new `-S` and `-B` arguments to explicitly specify source and build directories
- the `FindSQLite3` module was added to find the SQLite v3.x library (_not_ implemented in this PR yet)
- the `cmake --build` tool `--target` parameter gained support for multiple targets. It now also has a short form `-t` alias
- the `cmake` command gained a new `--install` option

### CODE REVIEW

**Comment by hebasto** [comment]
cc @fanquake @theuni @TheCharlatan @luke-jr @willcl-ark @stickies-v @ajtowns @w0xlt @pablomartin4btc

**Comment by ajtowns** [comment]
seems sensible to me

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,20 +5,20 @@
...
+#  - CMake 3.18.4, https://packages.debian.org/buster-backports/cmake
 #
-# Ubuntu 22.04 Jammy, https://wiki.ubuntu.com/Releases, EOL 2032:
-#  - CMake 3.22.1, https://packages.ubuntu.com/jammy/cmake
+# Ubuntu 20.04 LTS Jammy, https://wiki.ubuntu.com/Releases, EOSS in April 2025:
+#  - CMake 3.16.3, https://packages.ubuntu.com/focal/cmake
 #
-# Visual Studio 17 2022, https://visualstudio.microsoft.com:
-#  - CMake 3.24
+# Ubuntu 22.04 LTS Jammy, https://wiki.ubuntu.com/Releases, EOSS in June 2027:
```
Comment (fanquake) [inline CMakeLists.txt:14]: Why are we listing multiple version of Ubuntu here. Seems unecessary to repeat things that are trivial to look up, and the only one that matters is our minimum supported?

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,20 +5,20 @@
...
 #            by CMake version.
 #
-# Debian 10 Buster, https://wiki.debian.org/LTS, EOL 2024:
-#  - CMake 3.13.4, https://packages.debian.org/buster/cmake
+# Debian 10 Buster, https://wiki.debian.org/LTS, EOL in June 2024:
+#  - CMake 3.18.4, https://packages.debian.org/buster-backports/cmake
 #
-# Ubuntu 22.04 Jammy, https://wiki.ubuntu.com/Releases, EOL 2032:
-#  - CMake 3.22.1, https://packages.ubuntu.com/jammy/cmake
+# Ubuntu 20.04 LTS Jammy, https://wiki.ubuntu.com/Releases, EOSS in April 2025:
```
[the remark this answers is not in the document]
Comment (hebasto) [reply CMakeLists.txt]: Thanks! Fixed.

File: CMakeLists.txt (hunk 1 above)
Comment (hebasto) [reply CMakeLists.txt]: Thanks! Reworked.

File: CMakeLists.txt [hunk 3]
```diff
@@ -166,19 +166,15 @@ add_boost_if_needed()
 
 include(CheckSourceCompilesAndLinks)
 
-if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.14)
-  include(CheckPIESupported)
-  check_pie_supported(OUTPUT_VARIABLE check_pie_output LANGUAGES CXX)
-  if(CMAKE_CXX_LINK_PIE_SUPPORTED)
-    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
-  endif()
+include(CheckPIESupported)
```
Comment (fanquake) [inline CMakeLists.txt:169]: In 66f1e5b766d071a53f45e1f8436d9c8d8cc717e1:

Not sure what's happening here, but this check doesn't work, and I don't get PIE.
```bash
CMake Warning at CMakeLists.txt:174 (message):
  PIE is not supported at link time: PIE (CXX): Change Dir:
...
  Position independent code disabled.
```

Full output here: https://gist.github.com/fanquake/1c38b5367a8cca924fd732e7b5fe7817.

File: CMakeLists.txt [hunk 4]
```diff
@@ -5,23 +5,17 @@
...
-# Debian 10 Buster, https://wiki.debian.org/LTS, EOL 2024:
-#  - CMake 3.13.4, https://packages.debian.org/buster/cmake
-#
-# Ubuntu 22.04 Jammy, https://wiki.ubuntu.com/Releases, EOL 2032:
-#  - CMake 3.22.1, https://packages.ubuntu.com/jammy/cmake
-#
-# Visual Studio 17 2022, https://visualstudio.microsoft.com:
-#  - CMake 3.24
+# Ubuntu 20.04 LTS Focal Fossa, https://wiki.ubuntu.com/Releases, EOSS in April 2025:
+#  - CMake 3.16.3, https://packages.ubuntu.com/focal/cmake
```
[the remark this answers is not in the document]
Comment (hebasto) [reply CMakeLists.txt:9]: Sure. Added.

**Review by TheCharlatan — APPROVED** [review]
ACK b517d5c145939667e5d773d0a9b5bc85ceb05a67

File: CMakeLists.txt [hunk 5]
```diff
@@ -77,6 +77,16 @@ set(CMAKE_CXX_EXTENSIONS OFF)
...
 
+include(CheckPIESupported)
+check_pie_supported(OUTPUT_VARIABLE check_pie_output LANGUAGES CXX)
+if(CMAKE_CXX_LINK_PIE_SUPPORTED)
+  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+else()
+  message(WARNING "PIE is not supported at link time: ${check_pie_output}")
+  list(APPEND configure_warnings "Position independent code disabled.")
+endif()
+unset(check_pie_output)
```
Comment (theuni) [inline CMakeLists.txt:88]: This `unset` is new in the rebase, no? Is it needed for some reason? Is this a pattern we should be reviewing for?

**Review by theuni — APPROVED** [review]
ACK b517d5c145939667e5d773d0a9b5bc85ceb05a67

File: CMakeLists.txt (hunk 5 above)
Comment (hebasto) [reply CMakeLists.txt:88]: The reason is to keep global namespace tidy.

**Comment by fanquake** [comment]
Can someone link to the discussion about the CMake PIE checks, their output, and how it differs from what we do with master?
I guess we are fine with what is being done here, but I didn't see the difference being discussed anywhere. (I also don't see anything PIE related in the hardening PR, to account for any difference.)

i.e, one example:
```bash
# Master

src/init.cpp -fPIE
bitcoind -fPIE -pie

# CMake

src/init.cpp -fPIC
bitcoind -fPIE
```

**Comment by hebasto** [comment]
> Can someone link to the discussion about the CMake PIE checks, their output, and how it differs from what we do with master? I guess we are fine with what is being done here, but I didn't see the difference being discussed anywhere. (I also don't see anything PIE related in the hardening PR, to account for any difference.)
> 
> i.e, one example on Fedora Linux:
> 
> ```shell
> # Master
> 
> src/init.cpp -fPIE
> bitcoind -fPIE -pie
> 
> # CMake
> 
> src/init.cpp -fPIC
> bitcoind -fPIE
> ```

From CMake docs:
- https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html
- https://cmake.org/cmake/help/latest/policy/CMP0083.html
- https://cmake.org/cmake/help/latest/module/CheckPIESupported.html

**Comment by fanquake** [comment]
Sure, I can lookup the docs, but I'm asking why have we changed this, does the difference matter, etc?

**Comment by fanquake** [comment]
In the docs I also see:
> Set to TRUE: flags to produce a position independent executable are passed to the linker step. For example -pie for GCC.

but that doesn't seem to be the case here? So is the check not working as intended, is our behaviour differing from the docs?

**Comment by hebasto** [comment]
Autotools:
```
$ file src/bitcoind
src/bitcoind: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9b7e56fa795f50af442bd6930d9b9b21e683403d, for GNU/Linux 3.2.0, with debug_info, not stripped
```

CMake:
```
$ file build/src/bitcoind 
build/src/bitcoind: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=a028648ba6ed1937ae5be5d826adbca412251245, for GNU/Linux 3.2.0, with debug_info, not stripped
```

**Comment by hebasto** [comment]
The link command for `bitcoind` does include `-pie`:
```
$ cat build/src/CMakeFiles/bitcoind.dir/link.txt
/usr/bin/c++ -O2 -g -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -fPIE -pie CMakeFiles/bitcoind.dir/bitcoind.cpp.o CMakeFiles/bitcoind.dir/init/bitcoind.cpp.o -o bitcoind  libbitcoin_node.a wallet/libbitcoin_wallet.a /usr/lib/x86_64-linux-gnu/libsqlite3.so /usr/lib/x86_64-linux-gnu/libdb_cxx.so libbitcoin_common.a ../libsecp256k1.a util/libbitcoin_util.a ../libminisketch.a /usr/lib/x86_64-linux-gnu/libevent_pthreads.so /usr/lib/x86_64-linux-gnu/libevent.so /usr/lib/x86_64-linux-gnu/libnatpmp.so /usr/lib/x86_64-linux-gnu/libminiupnpc.so zmq/libbitcoin_zmq.a ../libleveldb.a ../libcrc32c.a univalue/libunivalue.a /usr/lib/x86_64-linux-gnu/libzmq.so 
```

**Comment by theuni** [comment]
CMake added the necessary linker behavior here: https://cmake.org/cmake/help/latest/policy/CMP0083.html

This is kinda like https://github.com/bitcoin/bitcoin/pull/28672 IMO. pic/pie are arguably used out of practicality/compatibility more these days than for hardening itself. I think it's reasonable to disconnect this from hardening.

**Comment by theuni** [comment]
Wait, no. I misunderstood what was happening here. I thought we agreed not to force this on, letting CMake do the right thing depending on what was being built? Wasn't this only turned on as a stop-gap for initial review?

**Comment by hebasto** [comment]
> Wait, no. I misunderstood what was happening here. I thought we agreed not to force this on, letting CMake do the right thing depending on what was being built? Wasn't this only turned on as a stop-gap for initial review?

Just to clarify, this PR did not aim to alter the PIC/PIE behavior. It dropped the code branch for CMake versions older than 3.16.

**Comment by theuni** [comment]
Yes, understood. And I think above I was misremembering, that's how we handled it for libsecp. For Core, yeah, just turn it on.

All good. Sorry for the noise.

**Comment by hebasto** [comment]
Speaking of PIC/PIE, I really doubt that the check, or at least a warning, is required for a Windows target (both MSVC and cross-compiling).

**Comment by hebasto** [comment]
>    * the `FindSQLite3` module was added to find the SQLite v3.x library (_not_ implemented in this PR yet)

Implemented in https://github.com/hebasto/bitcoin/pull/39.

**Pull request merged**